### ISSUE TITLE
aws eks update-kubeconfig invalid apiVersion

### ISSUE THREAD

**Issue by code-is-art** [aws/aws-cli#6920]
### Describe the bug

Update kubectl from v1.23.6 to 1.24.0 and run commands
```bash
aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}
kubectl version
```
I get the following error and exit status 1
```
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
```
Kubectl must need an updated apiVersion in kubeconfig file. Not sure if this is on aws side or kubectl side.
https://github.com/kubernetes/kubectl/issues/1210

### Expected Behavior

No error message when using `kubectl` and `aws eks update-kubeconfig`

### Current Behavior

error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"

### Reproduction Steps

Update kubectl from v1.23.6 to 1.24.0 and run commands
```bash
aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}
kubectl version
```

### CLI version used

whatever is running in aws/codebuild/standard:5.0

### Environment details (OS name and version, etc.)

aws/codebuild/standard:5.0

**Comment by tim-finnigan** [comment]
Hi @code-is-art thanks for reporting. We just received a feature request related to this: https://github.com/aws/aws-cli/issues/6916. 

There is also a related PR open (https://github.com/aws/aws-cli/pull/6476) that the team is going to prioritize reviewing. 

Are you able to pin to a previous version of Kubernetes as a temporary solution?

**Comment by code-is-art** [comment]
yes v1.23.6 works fine. I modified my pipelines to stay at this version for now. I was doing
```bash
curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl
```
but now I am doing the following as a solution.
```bash
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/linux/amd64/kubectl
```
I believe this version of kubectl was released yesterday and my new pipeline picked it up.

**Comment by Zheaoli** [comment]
https://gist.github.com/Zheaoli/335bba0ad0e49a214c61cbaaa1b20306

Here's the workround

**Comment by bpourian** [comment]
If you are still having issues in your pipeline or after you install make sure to check the version:
`curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/linux/amd64/kubectl
 mv ./kubectl /usr/local/bin/kubectl`

**Comment by kyandaks** [comment]
Still experiencing the same issue.
I am using GitHub Runners and the kubectl version on the runners is 1.24.0. 
Still figuring out a way to update this to a version less than 1.24.0

**Comment by genieai-vikas** [comment]
Facing same issue

**Comment by yermukov** [comment]
same here

**Comment by beeshalrizal** [comment]
same here

**Comment by coman-andrei** [comment]
same here

**Comment by code-is-art** [comment]
There is a thumbs up button for the` same here` crowd. Unless you have something to add to the conversation please use it. No one like to be spammed with same here emails and notifications.

**Comment by 1saf** [comment]
same here

**Comment by TranToan2910** [comment]
I have same problem with github runner.
I switched back to version 1.23.6 using this action and it works now
https://github.com/Azure/setup-kubectl

`
      - uses: azure/setup-kubectl@v2.0
        with:
          version: 'v1.23.6'
        id: install
`

**Comment by filipepratalima** [comment]
Same thing here, even simply getting the kubectl version is showing this error:
`➜  ~ sudo apt search kubectl
Sorting... Done
Full Text Search... Done
kubectl/kubernetes-xenial,now 1.24.0-00 arm64 [installed]
  Kubernetes Command Line Tool

kubetail/focal 1.6.5-2 all
  Aggregate logs from multiple Kubernetes pods into one stream

➜  ~ kubectl version        
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"`

**Comment by David-Tamrazov** [comment]
> 

worked like a charm for us; thank you @TranToan2910 ! ❤️

**Comment by FernandoAlencarJr** [comment]
same problem when using github actions ...

**Comment by SlaviDimitrov** [comment]
we also have the same problem when using GitHub Actions

**Comment by tim-finnigan** [comment]
Thanks all for your feedback. As someone mentioned earlier, please 👍 the original post if you are affected by this and only leave a comment if you have additional information to share.

Kubernetes v1.24.0 was released Tuesday, May 3rd according to here: https://kubernetes.io/releases/

Several users have reported this breaking their workflows, many of which involve the use of GitHub Actions/Runners. The team is actively looking into this further and I will share an update as soon as I have more information.

In the meantime, reverting to v1.23.6 is a temporary solution as mentioned in comments [here](https://github.com/aws/aws-cli/issues/6920#issuecomment-1117981158) and [here](https://github.com/aws/aws-cli/issues/6920#issuecomment-1118569100).

**Comment by jcs47** [comment]
> I have the same problem with GitHub runner.
> I switched back to version 1.23.6 using this action and it works now
> https://github.com/Azure/setup-kubectl

I am having issues with that action:

```
Error: Error: Kubectl '1.23.6' for 'amd64' arch not found.
```

**Comment by lasanthak** [comment]
For MAC:

First verify the version:
$ kubectl version --client=true

Download and install 1.23.6:
https://dl.k8s.io/release/v1.23.6/bin/darwin/amd64/kubectl

Fixed!

**Comment by bkershaw** [comment]
Our EKS cluster is running 1.21 currently and the github action just failed with error:
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"

This is github action is an "apply -f" command" not "update-kubeconfig".

**Comment by qgxpagamentos** [comment]
``> Our EKS cluster is running 1.21 currently and the github action just failed with error: error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
> 
> This is github action is an "apply -f" command" not "update-kubeconfig".

I went through the same problem!
Ubuntu from Github has version 1.24.0 installed, it was necessary to do a downgrande!

I created an Action that you define which version will be on the machine and it worked for me! :smiley:

```yaml
on: push
name: deploy
jobs:
  deploy:
    name: deploy to cluster
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Install kubectl
        uses: qgxpagamentos/kubectl-aws-args@v1
        with:
          base64-kube-config: ${{ secrets.KUBE_CONFIG_DATA }}
          kubectl-version: v1.22.5

      - name: Deploy Kubernetes cluster Service/Deployment
        run: |
          kubectl version
          kubectl apply -f deployment/kubernetes
```

**Comment by kyleknap** [comment]
Hi everyone. I just wanted to give an update here. To make the `eks get-token` and `eks update-kubeconfig` commands compatible with kubectl 1.24.0, we are planning on merging this PR: https://github.com/aws/aws-cli/pull/6476. Specifically this change will:

* If available, the `eks get-token` command retrieve and return the API version from the `KUBERNETES_EXEC_INFO` environment variable that is set by `kubectl` when it calls out to `eks get-token` for authentication. The environment variable is the [officially documented way](https://kubernetes.io/docs/reference/_print/#input-and-output-formats) to determine the API version to use. If the environment variables is not present, the command will continue to default to v1alpha1.

* Update the default API version set by `eks update-kubeconfig` in the `~/.kube/config` file to v1beta1. The API version set in the `KUBERNETES_EXEC_INFO` is sourced from the `~/.kube/config`. So even after upgrading the CLI once this PR is merged, you will have to re-run `eks update-kubeconfig` so that you are configured to use v1beta1, since v1alpha1 is no longer supported by `kubectl` and you will likely be previously configured to use v1alpha1  in your  `~/.kube/config`.

The current plan is to merge this PR tomorrow and include it in a release of both v1 and v2. In the meantime, please give us feedback by:

* Add the :+1: reaction to this comment if you agree with this approach.
* Pull down the branch in the PR: https://github.com/aws/aws-cli/pull/6476 and confirm that this resolves compatibility issues you were seeing with 1.24.0. If it does fix the compatibility issue, add a 🚀  reaction to this comment. If it does not resolve the compatibility issues, please comment in this issue.
* Add a comment to this issue if you have alternative suggestions to the approach.

Outside of the compatibility issues, there are other improvements that are still in consideration outside of this change:

* Explicit setting of `--api-version` in the `eks get-token` and `eks update-kubeconfig` commands: https://github.com/aws/aws-cli/issues/6916
* Default `eks update-kubeconfig` to set API version to v1: https://github.com/aws/aws-cli/issues/6923

Let's continue any discussions for those improvements in their respective GitHub issues, but we do not see them required to unblock the current compatibility issues for 1.24.0. We recommend reacting with :+1: on those GitHub issues if you'd like to see them implemented.

**Comment by russellraed** [comment]
For those who are using circle ci and using aws-eks orb, this worked for me:
```
      - checkout
      - kubernetes/install-kubectl: # <- added
          kubectl-version: v1.23.6 # <- added
      - aws-eks/update-kubeconfig-with-authenticator:
          cluster-name: ${EKS_CLUSTER_NAME}
          aws-region: ${AWS_DEFAULT_REGION}
          install-kubectl: false # <- changed from [true] to [false]
          kubeconfig-file-path: ~/.kube/config
```
make sure you have the `kubernetes: circleci/kubernetes@1.3.0` orb added as well

**Comment by armujahid** [comment]
On debian/Ubuntu, I downgraded kubectl to previous version using `sudo apt install kubectl=1.23.6-00`

**Comment by mmdii** [comment]
> On debian/Ubuntu, I downgraded kubectl to previous version using `sudo apt install kubectl=1.23.6-00`

this one works for me thanks.

**Comment by lukaszraczylo** [comment]
Just in case anyone is looking for a oneliner solution working with the github actions

`sudo apt-get install -y curl awscli && curl -LO https://dl.k8s.io/release/v1.23.6/bin/linux/amd64/kubectl && sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl`

**Comment by maxsxu** [comment]
Same issue here.

- `kubectl v1.24.0`
- `aws-cli/2.6.2`

**Comment by bkershaw** [comment]
*Working Temporary Solution*
    - example snippet below

> Just in case anyone is looking for a oneliner solution working with the github actions
> 
> `sudo apt-get install -y curl awscli && curl -LO https://dl.k8s.io/release/v1.23.6/bin/linux/amd64/kubectl && sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl`

      - name: Check AWS version
        run: |
          aws --version
          aws configure set aws_access_key_id ${{secrets.AWS_ACCESS_KEY_ID}}
          aws configure set aws_secret_access_key ${{secrets.AWS_SECRET_ACCESS_KEY}}
          aws configure set region ${{secrets.EKS_REGION}}
          aws sts get-caller-identity

      - name: downgrade kubectl version (TEMPORARY SOLUTION)
        run: |
          sudo apt-get install -y curl awscli && curl -LO https://dl.k8s.io/release/v1.23.6/bin/linux/amd64/kubectl && sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl

      - name: Connect to EKS cluster 
        run: aws eks --region ${{secrets.EKS_REGION}} update-kubeconfig --name ${{secrets.EKS_CLUSTER}}

**Comment by divaypandey** [comment]
> ``> Our EKS cluster is running 1.21 currently and the github action just failed with error: error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
> 
> > This is github action is an "apply -f" command" not "update-kubeconfig".
> 
> I went through the same problem! Ubuntu from Github has version 1.24.0 installed, it was necessary to do a downgrande!
> 
> I created an Action that you define which version will be on the machine and it worked for me! 😃
> 
> ```yaml
> on: push
> name: deploy
> jobs:
>   deploy:
>     name: deploy to cluster
>     runs-on: ubuntu-latest
>     steps:
>       - name: Checkout
>         uses: actions/checkout@v2
> 
>       - name: Install kubectl
>         uses: qgxpagamentos/kubectl-set@v1
>         with:
>           base64-kube-config: ${{ secrets.KUBE_CONFIG_DATA }}
>           kubectl-version: v1.22.5
> 
>       - name: Deploy Kubernetes cluster Service/Deployment
>         run: |
>           kubectl version
>           kubectl apply -f deployment/kubernetes
> ```

the ONLY action that worked for me! the Azure one kept giving the AMD64 issue while others didn't correctly copy the config data

**Comment by tim-finnigan** [comment]
https://github.com/aws/aws-cli/pull/6476 was merged into the latest release so updating the CLI should resolve this. If anyone still experiences issues with this after updating then please let us know. Also please note the related feature requests mentioned in an [earlier comment](https://github.com/aws/aws-cli/issues/6920#issuecomment-1119141939) if you wish to contribute to the discussions in those issues. Thanks!

**Comment by ooolox** [comment]
Had this error today.
**What I did**

I'm on windows, so I simply typed this on the cmd
`
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/windows/amd64/kubectl.exe
`
And add it to PATH.

Alternatively,

Linux
```
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/linux/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl
```

macOS
```
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.5.2/bin/darwin/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl
```

**Comment by doublespaces** [comment]
>   - name: Downgrade kubectl version (TEMPORARY SOLUTION)
>     run: |
>       sudo apt-get install -y curl awscli && curl -LO https://dl.k8s.io/release/v1.23.6/bin/linux/amd64/kubectl && sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl


This worked for me, thank you.

**Comment by darvanen** [comment]
For anyone landing here looking for a solution with regards to EKS, don't downgrade your kubeconfig version locally, upgrade it on your cluster:

```
aws eks update-kubeconfig --name [your-cluster-name]
```

**Comment by aliusmiles** [comment]
@dallin-shogun would you mind sharing your kubectl/awscli versions and truncated kubeconfig?

I just can't make it work in any variations, driving me crazy.

kubectl 1.24.0, awscli 2.6.3, recreated kubeconfig - kubectl still trowing an error "error: You must be logged in to the server (the server has asked for the client to provide credentials)"


`aws eks get-token` still returns v1alpha1, tried `KUBERNETES_EXEC_INFO='{"apiVersion": "client.authentication.k8s.io/v1beta1"}'` and got get-token to return v1beta1 but kubectl still not logging in.

**Comment by aliusmiles** [comment]
argh, nevermind - forgot to add --role-arn while running update-kubeconfig 🤦🏼‍♂️

Fun fact, EKS is still on 1.22 so with kubectl 1.24 it warns exceeding max skew. ` brew unlink kubernetes-cli &&  brew link kubernetes-cli@1.22` to the rescue.

**Comment by jackdpeterson** [comment]
Just commenting for those on ubuntu and have kubectl installed via snap. A working fix for me is to:
```bash
snap switch kubectl --channel=1.23/stable && snap refresh
```

**Comment by danielmetta** [comment]
Hello everyone,

I'm the same problem, but I change version 1.24.0 to 1.23.6 and resolved my problem.

But I have a new problem, when I use the comand `kubectl logs name_of_pod` my result is `Error from server (InternalError): Internal error occurred: Authorization error (user=kube-apiserver-kubelet-client, verb=get, resource=nodes, subresource=proxy)`. Anyone with this problem?

Other problem is when use `kubectl apply -f KubeDeploy.yaml` my message is successful, but my image not working.

**Comment by cydRabbit** [comment]
@teochenglim 2a works for me. thx. using ubuntu on wsl2, aws cli 2.6.3, kubectl 1.24.0

**Comment by teochenglim** [comment]
@robinelvin try update the kubectl also to the latest version? Ours is 1.24.0

```shell
brew upgrade kubectl
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-03T13:36:49Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.9-eks-0d102a7", GitCommit:"eb09fc479c1b2bfcc35c47416efb36f1b9052d58", GitTreeState:"clean", BuildDate:"2022-02-17T16:36:28Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.24) and server (1.21) exceeds the supported minor version skew of +/-1
```

**Comment by teochenglim** [comment]
@robinelvin  maybe install your kubectl using homebrew? and not the downloaded binary?

https://docs.brew.sh/Installation
https://kubernetes.io/docs/tasks/tools/install-kubectl-macos/#install-with-homebrew-on-macos

```shell
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
brew install kubectl

### check with kubectl you are using, it is /usr/local/bin/kubectl ?
$ which kubectl                   ### is this from /usr/local/bin/kubectl ?
$  ls -l /usr/local/bin/kubectl ## is this from homebrew?
lrwxr-xr-x  1 chenglim.teo  admin  43 May 11 09:20 /usr/local/bin/kubectl -> ../Cellar/kubernetes-cli/1.24.0/bin/kubectl

### if needed
brew link --overwrite kubernetes-cli

```

**Comment by takarabt** [comment]
Hi everyone,

I did the steps above:
1- Update aws cli
2- update kub config

But still getting the error:

`Unable to connect to the server: getting credentials: decoding stdout: no kind "ExecCredential" is registered for version "client.authentication.k8s.io/v1alpha1" in scheme "pkg/client/auth/exec/exec.go:62"`

I'm using Docker desktop for windows. 
Client Version: v1.24.0
Kustomize Version: v4.5.4

**Comment by danielmetta** [comment]
I have two kubernetes, the first use version 1.20 in this version I can use the command `kubectl logs name_of_pod`.

But when I use this same command `kubectl logs name_of_pod` in my second kubernetes version 1.22 always returns this message `Error from server (InternalError): Internal error occurred: Authorization error (user=kube-apiserver-kubelet-client, verb=get, resource=nodes, subresource=proxy)`

When I change my kubectl to version 1.24 I have the problem `error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"` and if a change my config file to `client.authentication.k8s.io/v1beta1` I received this message `Unable to connect to the server: getting credentials: exec plugin is configured to use API version client.authentication.k8s.io/v1beta1, plugin returned version client.authentication.k8s.io/v1alpha1`

I don't know what to do now, because I already changed the version to 1.24, I already went back to 1.23, and always the same errors. What's different about this kubernetes 1.22.

I use Windows 11, my kubectl version now is `Client Version: v1.22.5`

Have there been any changes in version 1.24 two days ago?

Tks

**Comment by tim-finnigan** [comment]
For those still having issues, make sure that you have [updated your CLI version](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) to at least **1.23.9** in [v1](https://github.com/aws/aws-cli/blob/develop/CHANGELOG.rst) or **2.6.3** in [v2](https://github.com/aws/aws-cli/blob/v2/CHANGELOG.rst).

You then may need to delete your `~/.kube/config` file and run `aws eks update-kubeconfig --name [cluster_name]` to recreate it as mentioned [here](https://github.com/aws/aws-cli/issues/6920#issuecomment-1123080982).

**Comment by vanphan24** [comment]
Hi, Im still having the same issues after follow [steps ](https://github.com/aws/aws-cli/issues/6920#issuecomment-1123958587):
- I tried to reinstall using brew. It is version 1.24: 
`kubectl version --short --client
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.24.0
Kustomize Version: v4.5.4
`
- I deleted ~/.kube/config 
- I re-ran aws eks update-kubeconfig --name <my-eks-cluster-name> --region <my-aws-region>

The above got same error:
`error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"`

- I also tried to change the ~/.kube/config file with apiVersion: client.authentication.k8s.io/v1beta1 but got 
`Unable to connect to the server: getting credentials: decoding stdout: no kind "ExecCredential" is registered for version "client.authentication.k8s.io/v1alpha1" in scheme "pkg/client/auth/exec/exec.go:62"`

Maybe I missed some other crucial step? Appreciate any help.

**Comment by vanphan24** [comment]
Update, ok so I updated `kubectl` using the [AWS steps](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) instead of using `brew` and it looks like it works now. Thank you all!

**Comment by takarabt** [comment]
I finally managed to make it working after updating again AWS CLI and kube config file using the command:

`aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}`

It basically created a new entry :

`clusters:
- cluster:
    certificate-authority-data: xxxx==
    server: https://xxxxx.eks.amazonaws.com
  name: xxxxxx-plane`

With a context and a user, the only difference with the old configuration I use aws-iam-authenticator but the new generated configuration uses aws command.

I will need to rewrite my kub config file to make the other clusters and namespaces working.

Thanks a lot for the help.

**Comment by Nurul9615** [comment]
I was able to fix this by upgrading my AWS CLI 2 on local Windows.

download the .msi file from Amazon website: https://awscli.amazonaws.com/AWSCLIV2.msi

use msiexec in command line to extract application from msi package:
msiexec /a %USERPROFILE%\Downloads\AWSCLIV2.msi /qb TARGETDIR=%USERPROFILE%\awscli

Now you can use awscli version 2 using the following command:
>%USERPROFILE%\awscliv2\Amazon\AWSCLIV2\aws --version

Just add this directory your PATH. 

Delete and re-recreate your kube config file.

Now connecting to EKS should work.

**Comment by nicorikken** [comment]
I got it working by updating BOTH AWS CLI 2 (2.0.26 -> 2.6.4) and kubectl ( -> v1.24.0). In my case the client was already updates via the automatically updating Ubuntu snap package.

**Comment by lasanthak** [comment]
I was able to get it working in osx by running `aws eks update-kubeconfig` again for my current EKS cluster.
`kubectl version --client=true` --> 1.24.0
`aws --version` --> 2.6.3

**Comment by ppshein** [comment]
I've tried both `kubectl 1.23.6` and `1.24.0`, unfortunately both are not working at all. And my aws cli version is `aws-cli/2.6.4 Python/3.9.11 Linux/4.14.248-189.473.amzn2.x86_64 exe/x86_64.amzn.2 prompt/off`

and encountered following error message

```
Empty KUBERNETES_EXEC_INFO, defaulting to client.authentication.k8s.io/v1alpha1. This is likely a bug in your Kubernetes client. Please update your Kubernetes client.
Error: Kubernetes cluster unreachable: Get https://xxxxxxxxxxx.gr7.ap-southeast-1.eks.amazonaws.com/version?timeout=32s: getting credentials: exec plugin is configured to use API version client.authentication.k8s.io/v1beta1, plugin returned version client.authentication.k8s.io/v1alpha1
```

**Comment by marv254** [comment]
For this issue ensure the kubectl version of the local machine is the same as the kubernetes version of the eks service,,, downgraded the one on my local machine to 1.22 from 1.24 now its working as a charm,,,

**Comment by boogi** [comment]
Just change (e.g. with sed) v1alpha1 to v1beta1:

```
diff ~/.kube/config ~/.kube/config-backup
691c691
<             apiVersion: client.authentication.k8s.io/v1beta1
---
>             apiVersion: client.authentication.k8s.io/v1alpha1
```

**Comment by alexander-pwatcher** [comment]
> Update, ok so I updated `kubectl` using the [AWS steps](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) instead of using `brew` and it looks like it works now. Thank you all!

After this i had to do: 
aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}

Thanks!

**Comment by philipjmurphy** [comment]
> > Update, ok so I updated `kubectl` using the [AWS steps](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) instead of using `brew` and it looks like it works now. Thank you all!
> 
> After this i had to do: aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}
> 
> Thanks!

I also had to uninstall the brew version `brew uninstall awscli`, so that it would pick up the `AWS steps` installed version.

**Comment by jfrylings-twilio** [comment]
`brew upgrade kubectl` and `brew upgrade awscli` worked for me without having to uninstall `awscli` and install with the above AWS steps. YMMV

**Comment by fryerj** [comment]
`brew upgrade kubectl` and `brew upgrade awscli` worked for me without having to uninstall `awscli` and install with the above AWS steps. YMMV

**Comment by BloodyIron** [comment]
It really blows me away that there's no repository for these packages served by Amazon to Linux distros. Downloading binaries and copying them into folders is a bad practice which should already be painfully obvious by now. Package managers are a secure supply-chain mechanism to provide sources of software and updates. Downloading binaries like this leads to them generally not getting updated as they get forgotten, amongst other security issues. Really hope Amazon switches to repository delivery for these tools instead.

**Comment by BloodyIron** [comment]
Okay so the solution that I've found that is actually sensible is to downgrade kubectl from 1.24.0 to 1.23.6

On Ubuntu 22.04 or 20.04:

`sudo apt install kubectl=1.23.6-00`

Quite frankly I am appauled that there is zero concern about software being installed outside of package managers as the "solution" to this problem. 1.24.0 clearly has breaking changes made to it that should be coming from a reliable supply-chain (repositories) and again the recommended solution in this thread is to install awscli manually, breaking the supply-chain for software that is repositories and package managers. Professionally and personally I completely disagree with this practice.

So yeah, downgraded kubectl and I was immediately able to start working with kubernetes again (I had to use the "v1alpha1" apiVersion declaration though, not "v1beta1").

**Comment by stewartmoreland** [comment]
This change has already been implemented in AWS CLI v2.7.

If you've updated to kubectl `v1.24.0` already, update your AWS-CLI to version `>= 2.7.0`  and re-run `aws eks update-kubeconfig` to update your kube config with the `client.authentication.k8s.io/v1beta1` API endpoint.

**Comment by BloodyIron** [comment]
> This change has already been implemented in AWS CLI v2.7.
> 
> If you've updated to kubectl `v1.24.0` already, update your AWS-CLI to version `>= 2.7.1` and re-run `aws eks update-kubeconfig` to update your kube config with the `client.authentication.k8s.io/v1beta1` API endpoint.
> 
> https://github.com/aws/aws-cli/blob/2d59a71b94af1486ad5e2e81040924f2a934c5d0/.changes/1.24.0.json#L4

There's no deb packaging or repositories for equivalent for awscli of that version. Hence downgrading.

**Comment by gegham1** [comment]
@tim-finnigan hey, I have the same problem and I'm using CodeBuild.  Should not CodeBuild provide the latest version of aws-cli out of the box? or I should manually install 1.23.9  or 2.6.3 ? (I did not install aws-cli manually.)

**Comment by wanieldilson** [comment]
> @tim-finnigan hey, I have the same problem and I'm using CodeBuild. Should not CodeBuild provide the latest version of aws-cli out of the box? or I should manually install 1.23.9 or 2.6.3 ? (I did not install aws-cli manually.)

Did you figure out a solution when using CodeBuild? I tried @boogi's sed suggestion which was worth a go but no luck.

**Comment by inoas** [comment]
Temp fix for mac:

1. uninstall kubectl forced (keep the dependencies depending on it)
2. https://migueldavid.eu/using-multiple-versions-of-kubectl-on-macos install 

```
brew install asdf
asdf plugin-add kubectl  
asdf install kubectl 1.23.6
asdf global kubectl 1.23.6
```

**Comment by inoas** [comment]
btw... this is an irresponsible devops mess some choices created.

**Comment by code-is-art** [comment]
> > @tim-finnigan hey, I have the same problem and I'm using CodeBuild. Should not CodeBuild provide the latest version of aws-cli out of the box? or I should manually install 1.23.9 or 2.6.3 ? (I did not install aws-cli manually.)
> 
> Did you figure out a solution when using CodeBuild? I tried @boogi's sed suggestion which was worth a go but no luck.

The only thing that is working for me on codebuild is the following.
```
phases:
  install: # Install AWS cli, kubectl (needed for Helm) and Helm
    commands:
      - apt install -y git python3
      - echo installing awscli
      - curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
      - unzip awscliv2.zip
      - sudo ./aws/install --update
      - echo Setting up kubectl...
      - curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl
      - chmod +x kubectl
      - mv ./kubectl /usr/local/bin/kubectl
  pre_build: # Add kubeconfig to access to EKS cluster
    commands:
      - echo Update kubeconfig...
      - aws --version
      - aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}
      - kubectl version
```
FYI the following does not work at all
```
- pip3 install awscli --upgrade --user # THIS DOES NOT WORK AT ALL
```

**Comment by GrantThompson1** [comment]
For CodeBuild users using aws/codebuild/amazonlinux2-x86_64-standard:3.0 (most recent):

I found the default AWS CLI was AWS CLI v1.22.x

After downgrading kubectl to v1.23.6 , I had to update from the default CLI in the CodeBuild environment...

CodeBuild command to uninstall aws v1:
       - rm -rf /root/.pyenv/shims/aws
       - curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
       - unzip awscliv2.zip
       - sudo ./aws/install
       - aws --version

**Comment by loera140** [comment]
> For CodeBuild users using aws/codebuild/amazonlinux2-x86_64-standard:3.0 (most recent):
> 
> I found the default AWS CLI was AWS CLI v1.22.x
> 
> After downgrading kubectl to v1.23.6 , I had to update from the default CLI in the CodeBuild environment...
> 
> CodeBuild command to uninstall aws v1: - rm -rf /root/.pyenv/shims/aws - curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" - unzip awscliv2.zip - sudo ./aws/install - aws --version

Worked! Thank you

**Comment by wanieldilson** [comment]
I added this script to our buildspec.yml in codepipeline, fixed it for me as per @GrantThompson1's suggestion.

```shell

#!/bin/bash

curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/linux/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl

sudo rm -rf /root/.pyenv/shims/aws
curl https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip -o awscliv2.zip
unzip awscliv2.zip
sudo ./aws/install
aws --version

```

**Comment by code-is-art** [comment]
For me updating to the latest awscli using the method below allowed me to use the latest kubectl 1.2.4
```
sudo rm -rf /root/.pyenv/shims/aws
curl https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip -o awscliv2.zip
unzip awscliv2.zip
sudo ./aws/install
aws --version
```

**Comment by BloodyIron** [comment]
> For CodeBuild users using aws/codebuild/amazonlinux2-x86_64-standard:3.0 (most recent):
> 
> I found the default AWS CLI was AWS CLI v1.22.x
> 
> After downgrading kubectl to v1.23.6 , I had to update from the default CLI in the CodeBuild environment...
> 
> CodeBuild command to uninstall aws v1: - rm -rf /root/.pyenv/shims/aws - curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" - unzip awscliv2.zip - sudo ./aws/install - aws --version

If ever we needed more examples of why package managers are a good idea, here it is.

I'm sure plenty of people appreciate these steps being spelled out, but if this whole situation was properly handled through package managers, repositories, and related supply chains, then we wouldn't have to deal with manual software management to this degree.

Anyone ever heard of Bonzi Buddy?

**Comment by pumu-98** [comment]
this works for me

**Comment by BloodyIron** [comment]
> Previously, using `sed -i 's|client.authentication.k8s.io/v1alpha1|client.authentication.k8s.io/v1beta1|g' ~/.kube/config` was working for me (even deployed successfully as early as this morning). However, within the past hour, this solution stopped working and I am now getting an error stating my client api version for k8s authentication does not match v1beta1.
> 
> I tried switching to the downgrade solution by adding @GrantThompson1 's solution to the prebuild stage of codebuild, but it fails here too saying that v1alpha1 is an invalid version.
> 
> So I am getting mixed signals here. On one hand, v1alpha1 is invalid, while v1beta1 is a mismatch with the client. This is a mess, is anyone experiencing the same thing and are there any good temporary solutions until this is sorted out?

I switched back to v1alpha1 and downgraded kubectl package to "1.23.6-00" (naming in Ubuntu). I honestly cannot see a reasonable way to get kubectl 1.24 operational (because downloading stand alone binaries is a really bad practice vs package managers).

Perhaps try downgrading your kubectl and switching back to v1alpha1?

**Comment by hagaiovadia** [comment]
Proper fix for Mac OSX users:
```
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/darwin/amd64/kubectl

 chmod +x ./kubectl

 sudo mv ./kubectl /usr/local/bin/kubectl
 ```

**Comment by hagay3** [comment]
Proper fix for Mac OSX users:
```
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/darwin/amd64/kubectl

chmod +x ./kubectl

sudo mv ./kubectl /usr/local/bin/kubectl
 ```

**Comment by Igor-Yavych** [comment]
What helped me in the end was 
`aws eks --region region update-kubeconfig --name cluster_name`
and all it did was change 
`apiVersion: client.authentication.k8s.io/v1alpha1` to  `apiVersion: client.authentication.k8s.io/v1beta1` in kube config, located at 
`C:\Users\YourUser\.kube\config` on Windows

**Comment by inoas** [comment]
> Proper fix for Mac OSX users:
> 
> ```
> curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/darwin/amd64/kubectl
> 
> chmod +x ./kubectl
> 
> sudo mv ./kubectl /usr/local/bin/kubectl
> ```

I feel the proper fix is to use asdf instead of brew package for kubectl as explained here: https://github.com/aws/aws-cli/issues/6920#issuecomment-1131850465

Then per project you can even switch kubectl around.

**Comment by hagay3** [comment]
> > Proper fix for Mac OSX users:
> > ```
> > curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/darwin/amd64/kubectl
> > 
> > chmod +x ./kubectl
> > 
> > sudo mv ./kubectl /usr/local/bin/kubectl
> > ```
> 
> I feel the proper fix is to use asdf instead of brew package for kubectl as explained here: [#6920 (comment)](https://github.com/aws/aws-cli/issues/6920#issuecomment-1131850465)
> 
> Then per project you can even switch kubectl around.

I prefer not install more package managers if i dont have to.

**Comment by inoas** [comment]
> I prefer not install more package managers if i dont have to.

Agreed, no need for `nvm`, `rvm` or `pyenv` if you can have `asdf` :)

**Comment by mrsiejas** [comment]
> Proper fix for Mac OSX users:
> 
> ```
> curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/darwin/amd64/kubectl
> 
> chmod +x ./kubectl
> 
> sudo mv ./kubectl /usr/local/bin/kubectl
> ```

I wouldn't call hardcoding specific version a `proper fix`. That could work as temporary workaround on servers but cumbersome and counterintuitive on workstations in a long run.

**Comment by norru** [comment]
I'm still getting this and using the downgrade workaround.

I see the issue has been closed but I can't find at which point it was actually fixed and what the fix is.

On top of that, there doesn't seem to be a `.deb` for `awscli 2`. Is the only way to install on Ubuntu extracting the tarballs?

**Comment by BloodyIron** [comment]
I can't fathom why the security and update implications of manual binary installation is being literally ignored here.

**Comment by BloodyIron** [comment]
> > Proper fix for Mac OSX users:
> > ```
> > curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/darwin/amd64/kubectl
> > 
> > chmod +x ./kubectl
> > 
> > sudo mv ./kubectl /usr/local/bin/kubectl
> > ```
> 
> I wouldn't call hardcoding specific version a `proper fix`. That could work as temporary workaround on servers but cumbersome and counterintuitive on workstations in a long run.

It's a bad idea even for servers. Because it will be easily forgotten and _NEVER UPDATED AGAIN_. That's the whole point of package managers _especially on servers_. Single source of truth for what needs to be updated, how to install, etc.

**Comment by dorianmariefr** [comment]
Using kubectl 1.21.9 fixed it for me, with asdf:

```
asdf plugin-add kubectl https://github.com/asdf-community/asdf-kubectl.git
asdf install kubectl 1.21.9
```

**Comment by brainwipe** [comment]
For me (dev machine, Windows 10), I updated the AWS CLI (to version 2.7.4) and re-ran the kube config command:

```
aws eks update-kubeconfig --name ${EKS_CLUSTER_NAME} --region ${REGION}
kubectl version
```

And it worked as expected. The change that was made was the use of `apiVersion` to `v1beta1`.

**Comment by iDataist** [comment]
After running `sed -i .bak -e 's/v1alpha1/v1beta1/' ~/.kube/config`, I got a new error. I'm not sure how to resolve it.

`kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.6", GitCommit:"ad3338546da947756e8a88aa6822e9c11e7eac22", GitTreeState:"clean", BuildDate:"2022-04-14T08:49:13Z", GoVersion:"go1.17.9", Compiler:"gc", Platform:"darwin/amd64"}
Unable to connect to the server: getting credentials: exec plugin is configured to use API version client.authentication.k8s.io/v1beta1, plugin returned version client.authentication.k8s.io/v1alpha1`

**Comment by Yurko-Fedoriv** [comment]
for Ubuntu 
sudo snap remove kubectl
sudo snap install kubectl --channel=1.23/stable --classic

**Comment by davyzhang3** [comment]
For M1 Mac users: 
`brew uninstall kubectl`
`brew install kubectl@1.22`
solved this issue for me

**Comment by samceena** [comment]
Manually installing the [aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator/releases) 0.5.6 worked for me. 
Also, it's worth noting that i'm running AWS EKS 1.21

**Comment by erickhun** [comment]
I had an old version of awscli (`2.6.2`). Upgrading it to the latest one `2.7.6` solved the issue

```
$ aws --version
aws-cli/2.6.2

brew upgrade awscli

$ aws --version
aws-cli/2.7.6

```

**Comment by sobi3ch** [comment]
`eks get-token` has changed in AWS CLI 2.7.x from `client.authentication.k8s.io/v1alpha1` -> `client.authentication.k8s.io/v1beta1`.
https://github.com/aws/aws-cli/blob/v2/CHANGELOG.rst#L109

```
$ aws --version
aws-cli/2.7.6 Python/3.9.11 Linux/5.4.0-113-generic exe/x86_64.ubuntu.20 prompt/off

$ aws eks get-token --cluster-name ockam-production | jq . | head -n4
{
  "kind": "ExecCredential",
  "apiVersion": "client.authentication.k8s.io/v1beta1",
  "spec": {},

$
```

**Comment by sobi3ch** [comment]
`eks get-token` has changed in AWS CLI 2.7.x from `client.authentication.k8s.io/v1alpha1` -> `client.authentication.k8s.io/v1beta1`.
https://github.com/aws/aws-cli/blob/v2/CHANGELOG.rst#L109

```
$ aws --version
aws-cli/2.7.6 Python/3.9.11 Linux/5.4.0-113-generic exe/x86_64.ubuntu.20 prompt/off

$ aws eks get-token --cluster-name cluster-name | jq . | head -n4
{
  "kind": "ExecCredential",
  "apiVersion": "client.authentication.k8s.io/v1beta1",
  "spec": {},

$
```

**Comment by sharadiiita** [comment]
> I had an old version of awscli (`2.6.2`). Upgrading it to the latest one `2.7.6` solved the issue
> 
> ```
> $ aws --version
> aws-cli/2.6.2
> 
> brew upgrade awscli
> 
> $ aws --version
> aws-cli/2.7.6
> ```

Nope I updated to the latest API version but still the same error on Mac

```
➜  ~ aws --version      
aws-cli/2.7.6 Python/3.9.13 Darwin/21.5.0 source/x86_64 prompt/off
➜  ~ kubectl get pods           
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
```

**Comment by mrliptontea** [comment]
@sharadiiita You need either older kubectl and aws CLI, or update both and update your `~/.kube/config` by either:
* replacing `client.authentication.k8s.io/v1alpha1` with `client.authentication.k8s.io/v1beta1`
* or running `aws eks update-kubeconfig --region ${REGION} --name ${EKS_CLUSTER_NAME}`

**Comment by ShradhaKhard** [comment]
I uninstalled the kubectl, kubectx and reverted back to the older version of kube client 
- brew uninstall kubectx
- brew uninstall --ignore-dependencies kubectl
- brew install kubernetes-cli@1.22
- Recreated my ~/.kube/config

and volla 

The CLI version for me:
`aws-cli/2.2.32 Python/3.8.8 Darwin/20.6.0 exe/x86_64 prompt/off`

**Comment by norru** [comment]
Manually installing (eurgh!) aws-cli version 2.7.7 from the zipball fixed the issue for me (I am now able to upgrade kubectl to 1.24.0).
The lack of a `.deb` package for Ubuntu is a major workflow issue though.

```
$ aws --version
aws-cli/2.7.7 Python/3.9.11 Linux/5.13.0-44-generic exe/x86_64.ubuntu.20 prompt/off
```

**Comment by nbcarey** [comment]
For what its worth, I extracted the HomeBrew formula for `kubernetes-cli` version `1.23.6` into its own HomeBrew tap at https://github.com/nbcarey/homebrew-tap-kubernetes-backlevel.

You should be able to install the back-level version by either

```shell
brew install nbcarey/tap-kubernetes-backlevel/kubernetes-cli@1.23.6
```

Or

```shell
brew tap nbcarey/tap-kubernetes-backlevel
brew install kubernetes-cli@1.23.6
```

**Comment by gaminfra** [comment]
I used mac M1  , I solved my issues .  

kubectl get nodes
NAME                                             STATUS   ROLES    AGE     VERSION
ip-10-1-10-10.ap-southeast-1.compute.internal    Ready    <none>   3d18h   v1.22.9-eks-810597c
ip-10-1-10-17.ap-southeast-1.compute.internal    Ready    <none>   36m     v1.22.9-eks-810597c
ip-10-1-10-234.ap-southeast-1.compute.internal   Ready    <none>   3d18h   v1.22.9-eks-810597c
ip-10-1-12-215.ap-southeast-1.compute.internal   Ready    <none>   36m     v1.22.9-eks-810597c

------------------------------------------------------------
1- replaced v1alpha1 to v1beta1 in ~/.kube/config file   

user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1 #client.authentication.k8s.io/v1beta1

2-reinstall aws cli 

brew install awscli  
aws --version                    
aws-cli/2.0.30 Python/3.7.4 Darwin/21.2.0 botocore/2.0.0dev34

**Comment by nathan100b** [comment]
I have the same issue . I solved this with 2 step 
1-Replace v1alpha1 to v1beta1 in kube config file ~/.kube/config 
 _user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1 

2- Reinstall aws cli 
 _brew install awscli_ 

  aws --version                    
  aws-cli/2.0.30 Python/3.7.4 Darwin/21.2.0 botocore/2.0.0dev34

3- Result it work well :)
kubectl get nodes
NAME                                             STATUS   ROLES    AGE     VERSION
ip-10-1-10-10.ap-southeast-1.compute.internal    Ready    <none>   3d19h   v1.22.9-eks-810597c
ip-10-1-10-17.ap-southeast-1.compute.internal    Ready    <none>   44m     v1.22.9-eks-810597c
ip-10-1-10-234.ap-southeast-1.compute.internal   Ready    <none>   3d19h   v1.22.9-eks-810597c
ip-10-1-12-215.ap-southeast-1.compute.internal   Ready    <none>   44m     v1.22.9-eks-810597c

**Comment by govind-bt** [comment]
I am still seeing error while downgrading kubectl. Any thoughts?

```
│ Error: Kubernetes cluster unreachable: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"

$ aws --version
aws-cli/2.7.9 Python/3.9.11 Linux/5.10.102.1-microsoft-standard-WSL2 exe/x86_64.ubuntu.20 prompt/off

$ kubectl version

Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.6", GitCommit:"ad3338546da947756e8a88aa6822e9c11e7eac22", GitTreeState:"clean", BuildDate:"2022-04-14T08:49:13Z", GoVersion:"go1.17.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by SadaGowda** [comment]
For anyone on using brew, installed kubectl previous version and updated to 1.24.x , you can just unistall kubectl,
with brew uninstall kubectl, the older version version will still remain and kubectl version-1 will be set

**Comment by rolandalla** [comment]
On MAC M1:
~/.kube/config

Change : **v1alpha1** to  **v1beta1**

**Comment by vijaychhipa** [comment]
The above, changing v1alpha1 to v1beta1  in ~/.kube/config   worked on any Intel Mac also, not just M1.

**Comment by elviento** [comment]
same issue here, this resolved it eg. rolling back to k9s `v0.25.18` [here](https://github.com/derailed/k9s/releases/tag/v0.25.18)

**Comment by bossjones** [comment]
Just providing one more example of how to revert to an earlier version of kubectl if need be: [asdf](https://asdf-vm.com/#/plugins-all) which is a nifty tool that acts as a version manager for a slew of common tools(think rbenv or pyenv but for anything). Easy to extend too via their plugin system if you see a tool that isn't supported yet.

When setup, all you need to run is:

```
asdf install kubectl 1.21.13
asdf global kubectl 1.21.13
```

And now:

```
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.13", GitCommit:"80ec6572b15ee0ed2e6efa97a4dcd30f57e68224", GitTreeState:"clean", BuildDate:"2022-05-24T12:40:44Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.13-eks-84b4fe6", GitCommit:"e1318dce57b3e319a2e3fecf343677d1c4d4aa75", GitTreeState:"clean", BuildDate:"2022-06-09T18:22:07Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

I'm on a m1 mac if that matters.

Hope this helps someone.

**Comment by vikasedu10** [comment]
I have mainly faced this issue due to terminal suggestion that it gives to install awscli (using `sudo apt install awscli`) whenever I try to use awscli commands. 

Try to remove awscli (using `sudo apt remove awscli`) and install using official guild [here](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).

Hope this will fix the issue!! If not, feel free to comment.

**Comment by myroslavmail** [comment]
> On MAC M1: ~/.kube/config
> 
> Change : **v1alpha1** to **v1beta1**

making that change, lead to success to but not from the first run. I've encountered with the issue: `Unable to connect to the server: getting credentials: decoding stdout: no kind "ExecCredential" is registered for version "client.authentication.k8s.io/v1alpha1" in scheme "pkg/client/auth/exec/exec.go:62"`

however it was fixed, after I've upgraded my aws-cli (yes, I'm AWS user) to the latest version, as described here: https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html

**Comment by matheushent** [comment]
For those who's still struggling with this (me included), I solved (on Mac 12.5) wiping out kubectl and aws cli form my system, then I ran:

```shell
brew install awscli # it installed 2.7.6
brew install kubectl # it installed 1.24.1
```

**Comment by nitrocode** [comment]
I use a shell function that generates my kubectl profiles.

Here is the smaller shell function `get-api-version()` that I use to calculate whether to use `v1alpha1` or `v1beta1` in case anyone finds it useful.

https://gist.github.com/nitrocode/715d262cf705b1bf1fa5930b88e795e3

**Comment by sb-daniel-an** [comment]
```
➜  /usr/local/bin/aws --version
aws-cli/2.5.8 Python/3.9.11 Darwin/21.3.0 exe/x86_64 prompt/off
➜  /opt/homebrew/bin/aws --version
aws-cli/2.7.28 Python/3.10.6 Darwin/21.3.0 source/arm64 prompt/off
```

```
➜  diff ~/.kube/config.by_aws.2.7 ~/.kube/config.by_aws.2.5
19c19
<       apiVersion: client.authentication.k8s.io/v1beta1       # by 2.7
---
>       apiVersion: client.authentication.k8s.io/v1alpha1      # by 2.5
```

Following `kubectl` calls work fine when the apiVersion is `v1beta1`

**Comment by arpitha-m-g** [comment]
This has worked for me !
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.23.6/bin/linux/amd64/kubectl 
 mv ./kubectl /usr/local/bin/kubectl
 sudo chmod  +x   /usr/local/bin/kubectl

**Comment by j4ys0n** [comment]
if you're using aws, make sure your aws cli is up to date.

**Comment by neerajjain92** [comment]
what worked for me is this https://github.com/pulumi/pulumi-kubernetes/issues/2092#issuecomment-1249513298

**Comment by dantman** [comment]
> if you're using aws, make sure your aws cli is up to date.

This was the issue for me and it's an issue many users may not realize.

This exec command requires a v2 version of the AWS CLI.

However if you install the AWS CLI with `apt install awscli` as many of us used to do, you get the v1 AWS CLI. v2 has completely different instructions and no apt package. And v1 works completely fine other than stuff like this, so you won't even notice you're massively out of date.

**Comment by Oxyaction** [comment]
updating aws cli version to latest and re-running `aws eks update-kubeconfig` worked for me

**Comment by Shreedeepa** [comment]
ubuntu@ip-172-31-43-134:~$ kubectl get nodes
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1" - i am getting this error

Also, i tried there commands and also upgrade the EKS cluster to 1.23, but still getting the same error
curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
unzip awscliv2.zip
sudo ./aws/install --update
aws eks update-kubeconfig --region ${AWS_REGION}  --name ${EKS_CLUSTER_NAME}

**Comment by abjinugu** [comment]
> aws eks update-kubeconfig

same here..

**Comment by jamesrgrinter** [comment]
You need to update the kubectl/config entry after updating the awscli, i.e.
```
$ kubectl config delete-context $cluster_name
$ aws eks update-kubeconfig --region="$aws_region" --name="$cluster_name" --alias="$cluster_name" --profile="$aws_profile"
```
(adjusting the eks update-kubeconfig line to suit your particular needs and configuration)

**Comment by KalebMills** [comment]
@jamesrgrinter - This is the right answer for me (on Ubuntu). I had the same problem on my M1 Mac, and this fixed it as well. I originally added my kubeconfig via eksctl, which is what was causing me the issue (unknown to me what the actual issue is).

**Comment by vanga** [comment]
> You need to update the kubectl/config entry after updating the awscli, i.e.
> 
> ```
> $ kubectl config delete-context $cluster_name
> $ aws eks update-kubeconfig --region="$aws_region" --name="$cluster_name" --alias="$cluster_name" --profile="$aws_profile"
> ```
> 
> (adjusting the eks update-kubeconfig line to suit your particular needs and configuration)

this worked for me as well on Mac M1 (make sure aws cli is installed propery via brew or pip3 depending on how it was installed)

**Comment by advissor** [comment]
Upgrading the AWS CLI to latest version helps on Mac m1

**Comment by kaihendry** [comment]
@advissor latest version = v2 cli?

v2 aws cli isn't supported on Alpine making this quite a problem https://github.com/aws/aws-cli/issues/4685

We need latest v1 aws cli to work :sweat:

**Comment by james-ransom** [comment]
Fixed it for me: 

curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl

**Comment by Turhral** [comment]
Hi there! I've encountered the same issue myself, on my Macbook Pro. It worked fine not too long ago, **but,** the only thing I can think of that could have messed this up is downloading the desktop version of Docker. I do remember that kubectl worked fine for me before.

A post on Stackoverflow mentioned something about it as well, and also talked about the PATHs which might be part of the issue. Couldn't solve it myself, so I hope this information might help.

**Comment by abdennour** [comment]
Seems like kubernetes v1.24.x is a broken change. Indeed, the best way is to downgrade your client here to something below 1.24.x ( 1.23.x at least)

**Comment by saibaldey** [comment]
Me too have faced the same issue.

`saibal.dey@XT3YH0XR6G ~ % kubectl version
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
saibal.dey@XT3YH0XR6G ~ % kubectl version --client
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.0", GitCommit:"b46a3f887ca979b1a5d14fd39cb1af43e7e5d12d", GitTreeState:"clean", BuildDate:"2022-12-08T19:51:43Z", GoVersion:"go1.19.4", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.7
saibal.dey@XT3YH0XR6G ~ %`

So the kubectl client version was 1.26 & upon checking my EKS cluster it was found to be 1.24. So there is a version mismatch across client & the server. Then upon bit of a research I came to know that > a v1.26 client can communicate with v1.25, v1.26, and v1.27 control planes, which has given me a clue what could be the issue.

Then I downgraded the version of my kubectl to 1.24 & updated the kubeconfig file & things found to be working.
`saibal.dey@XT3YH0XR6G ~ % kubectl version --client
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-03T13:46:05Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.4
saibal.dey@XT3YH0XR6G ~ % aws eks update-kubeconfig --name platform-pocs
Updated context arn:aws:eks:ap-south-1:549942532493:cluster/platform-pocs in /Users/saibal.dey/.kube/config
saibal.dey@XT3YH0XR6G ~ % kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-03T13:46:05Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.8-eks-ffeb93d", GitCommit:"abb98ec0631dfe573ec5eae40dc48fd8f2017424", GitTreeState:"clean", BuildDate:"2022-11-29T18:45:03Z", GoVersion:"go1.18.8", Compiler:"gc", Platform:"linux/amd64"}
saibal.dey@XT3YH0XR6G ~ % kubectl get nodes
NAME                                            STATUS   ROLES    AGE   VERSION
ip-192-168-22-185.ap-south-1.compute.internal   Ready    <none>   26h   v1.24.7-eks-fb459a0
ip-192-168-76-152.ap-south-1.compute.internal   Ready    <none>   26h   v1.24.7-eks-fb459a0
saibal.dey@XT3YH0XR6G ~ %`

So, if you are facing the same issue across, please act accordingly.

**Comment by yalattas** [comment]
facing same issue

after performing `aws eks update-kubeconfig --name CLUSTER_NAME --region REGION

```
[ec2-user@ip-10-0-1-135 ~]$ kubectl get all
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
```

**Comment by yogendratamang48** [comment]
´Getting similar issue with EKS 1.25
```
$ kubectl version
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
$ kubectl version --client
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"8f94681cd294aa8cfd3407b8191f6c70214973a4", GitTreeState:"clean", BuildDate:"2023-01-18T15:58:16Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
```

**Comment by mcandre** [comment]
This is fixed in the latest awscli (v2 series) releases.

Mind, certain OS distributions continue to package the ancient awscli v1 series, which doesn't even authenticate properly with AWS, let alone EKS.

**Comment by ameyapagore** [comment]
@mcandre I installed the latest version on awscli (v2 series), still getting same issue:

`$ kubectl get all --kubeconfig "config"
error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
$ aws --version
aws-cli/2.10.3 Python/3.9.11 Linux/3.10.0-1160.36.2.el7.x86_64 exe/x86_64.centos.7 prompt/off
$ helm version
version.BuildInfo{Version:"v3.11.1", GitCommit:"293b50c65d4d56187cd4e2f390f0ada46b4c4737", GitTreeState:"clean", GoVersion:"go1.18.10"}
$ kubectl version --client
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"8f94681cd294aa8cfd3407b8191f6c70214973a4", GitTreeState:"clean", BuildDate:"2023-01-18T15:58:16Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7`

**Comment by praveenperera** [comment]
@mcandre not working

```
aws-cli/2.10.4 Python/3.11.2 Darwin/22.3.0 source/arm64 prompt/off
```

**Comment by StevenDewey** [comment]
@ameyapagore I resolved this by editing the `~/.kube/config` file.

Change all occurrences of `v1alpha1` to `v1beta1`.

Not sure if it is an excellent long-term solution, but it is working now!

**Comment by pglezen** [comment]
> @ameyapagore I resolved this by editing the `~/.kube/config` file.
> 
> Change all occurrences of `v1alpha1` to `v1beta1`.

I noticed this difference in `config` after I updated my AWS client and re-ran the `update-kubeconfig`.  Now my AWS CLI version is

```
$ aws --version
aws-cli/2.11.2 Python/3.11.2 Darwin/21.6.0 exe/x86_64 prompt/off
```

Also using `kubectl` version `Client Version: v1.26.1`.

**Comment by johnzheng1975** [comment]
Upgrade aws client version!!!!!!
This works!!!!!

**Comment by Kejzin** [comment]
I have the samme issue:
`error: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"`

My aws version:

`aws-cli/2.11.4 Python/3.11.2 Linux/5.15.0-67-generic exe/x86_64.ubuntu.22 prompt/off`

And I'm using nevest to date kubectl
`Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:40:17Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7`

when i switched v1alpha1 to v1beta1 it worked but is it a good workaround?

**Comment by rahulbaid** [comment]
For MacOS users still facing this issue,

I upgraded the aws CLI version (from `2.4.6` to `2.11.4`). If this `brew upgrade awscli` does not work, try this:
```
aws --version
rm '/usr/local/bin/aws'
brew install awscli
brew link --overwrite awscli
aws --version
aws configure sso
aws sso login
kubectl config delete-context ${CLUSTER_NAME}
```
and now copy-paste the AWS environment variables (short-term credentials) into the terminal session
```
aws sts get-caller-identity
aws eks update-kubeconfig --name ${CLUSTER_NAME} --region ${REGION_NAME}
kubectl version
```

**Comment by Vadim-Zenin** [comment]
Check kubectl config file:
`kubectl config view | grep 'client.authentication.k8s.io'`
You could see
`      apiVersion: client.authentication.k8s.io/v1alpha1`

Copy current kubectl config
````
if [[ -f ${HOME}/.kube/config ]]; then
    printf "INFO: copy ${HOME}/.kube/config file\n"
    cp ${HOME}/.kube/config ${HOME}/.kube/config.$(date +%Y%m%d-%H%M)
fi
````
Reconfigure kubectl
````
AWS_DEFAULT_REGION="${AWS_DEFAULT_REGION:-eu-west-1}"
AWS_PROFILE="${AWS_PROFILE:-<replace_me>}"
EKS_NAME="<replace_me>"
aws eks list-clusters --profile ${AWS_PROFILE} --region ${AWS_DEFAULT_REGION}
aws eks update-kubeconfig --name ${EKS_NAME} --profile ${AWS_PROFILE} --region ${AWS_DEFAULT_REGION}
aws eks describe-cluster --name ${EKS_NAME} --query cluster.status --profile ${AWS_PROFILE} --region ${AWS_DEFAULT_REGION}
````
Check if the issue exists:
`kubectl version`
Client Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.17-eks-a59e1f0", GitCommit:"567b6a1a716729b7a8c4196cf3b61d3c84df88f3", GitTreeState:"clean", BuildDate:"2023-03-09T19:54:56Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.10-eks-48e63af", GitCommit:"9176fb99b52f8d5ff73d67fea27f3a638f679f8a", GitTreeState:"clean", BuildDate:"2023-01-24T19:17:48Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
`aws --version`
aws-cli/2.11.11 Python/3.11.2 Linux/5.4.0-137-generic exe/x86_64.ubuntu.20 prompt/off

**Comment by NBrendel** [comment]
Got it working on `Ubuntu 22.04` by downgrading the `kubectl` version to 1.23.17 (installed via snap) 

`sudo snap install kubectl --channel=1.23/stable --classic`

**Comment by mballoni** [comment]
Same issue here. The min working version was 1.23 as well.

**Comment by vbarba88** [comment]
Worked by upgrading the aws cli version: from 2.2.41 to 2.11.9

**Comment by ikersuen** [comment]
Same issue still exist when using 

kubectl client version: v1.24.x
kubectl server version; v1.24.x

Still requires to downgrade to use client version v1.23.x for resolving the problem.

**Comment by arhimondr** [comment]
Thanks @rahulbaid , worked for me

**Comment by ratkokorlevski** [comment]
Worked by upgrading the aws cli version: from 2.2.41 to 2.11.9

**Comment by rmsys** [comment]
Thanks @rahulbaid! Worked for me, executing the second step session.

```
Ubuntu 20.04.6 LTS
aws-cli/2.13.0
kubectl v1.27.3
```

**Comment by charlesxjq** [comment]
Try to update the kubeconfig with changing "apiVersion: client.authentication.k8s.io/v1alpha1" into "apiVersion: client.authentication.k8s.io/v1beta1" and it works for me. I think maybe it causes by the upgrade of the kubernetes apiVersion.

**Comment by huitingyeong** [comment]
same issue. This worked for me:

- Updating AWSCLI from 2.4.9 to 2.13.19
- Run aws eks update-kubeconfig --region XXX --name XXX (this essentially updates ~/.kube/config and changes apiVersion to "client.authentication.k8s.io/v1beta1"

**Comment by pallavikamboj123** [comment]
> same issue. This worked for me:
> 
> * Updating AWSCLI from 2.4.9 to 2.13.19
> * Run aws eks update-kubeconfig --region XXX --name XXX (this essentially updates ~/.kube/config and changes apiVersion to "client.authentication.k8s.io/v1beta1"

Thankyou so much. This worked for me.

**Comment by lccoronel** [comment]
> same issue. This worked for me:
> 
> * Updating AWSCLI from 2.4.9 to 2.13.19
> * Run aws eks update-kubeconfig --region XXX --name XXX (this essentially updates ~/.kube/config and changes apiVersion to "client.authentication.k8s.io/v1beta1"

This worked for me.

**Comment by ulidtko** [comment]
But what about **awscli v1** ?

Since #4947 is still unresolved, many environments **cannot install**, much less update, awscli v2. Because `pip install "awscli >= 2"` does not work.

No EKS for those, right? :wink:

**Comment by ulidtko** [comment]
With:
 * awscli v1 — `aws-cli/1.22.34 Python/3.10.12 Linux/5.15.0-91-generic botocore/1.27.67`
 * EKS cluster version 21 — `cluster.platformVersion: eks.21` (in `aws eks describe-cluster`)
 * `v1alpha1` get-token apiVersion — `client.authentication.k8s.io/v1alpha1`

— I got it working by **downgrading kubectl** to latest available version before 1.24.0 (kubectl dropped `v1alpha1` in 1.24.0).

I.e. in Ubuntu, in my case, the latest version before 1.24.0 was 1.23.17:

```
sudo apt install kubectl=1.23.17-00
sudo apt-mark hold kubectl
```

`hold` disables further auth-updates.

**Comment by anhnguyenviet109** [comment]
> same issue. This worked for me:
> 
>     * Updating AWSCLI from 2.4.9 to 2.13.19
> 
>     * Run aws eks update-kubeconfig --region XXX --name XXX (this essentially updates ~/.kube/config and changes apiVersion to "client.authentication.k8s.io/v1beta1"

This works for me

**Comment by ozbillwang** [comment]
> Upgrade aws client version!!!!!! This works!!!!!

Yes, this is the correct solution; the other answers are rubbish. 😊

After uninstall aws cli v1 and install v2, it works

**Comment by gourav-rcrm** [comment]
Same here

**Comment by jacopofar** [comment]
I use `aws-iam-authenticator` and found out it has to be updated as well. Otherwise, even after updating `kubectl`, `aws-cli` and the version in the `~/.kube/config` file as indicated the error persist. I got the latest stable fresh from the [release](https://github.com/kubernetes-sigs/aws-iam-authenticator/releases) page and it worked again.
